print_nanny_webapp/subscriptions/views.py

```diff
@@ -0,0 +1,142 @@
+from print_nanny_webapp.subscriptions.forms import SubscriptionsPaymentForm
+import stripe
+import json, logging
+from django.conf import settings
+from django.http import HttpResponse, HttpRequest
+from django.urls import reverse
+from django.template.response import TemplateResponse
+
+import djstripe.models
+import djstripe.settings
+
+from print_nanny_webapp.dashboard.views import DashboardView
+
+logger = logging.getLogger(__name__)
+
+class SubscriptionsListView(DashboardView):
+    template_name = "subscriptions/list.html"
+    form_classes = {
+        "payment": SubscriptionsPaymentForm,
+    }
+
+    def get_context_data(self, *args, **kwargs):
+        ctx = super().get_context_data(*args, **kwargs)
+        ctx["STRIPE_PUBLIC_KEY"] = settings.STRIPE_PUBLIC_KEY
+        return ctx
+
+    def form_valid(self, form):
+        print("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")
+        stripe_source = form.cleaned_data["stripe_source"]
+        plan = form.cleaned_data["plan"]
+
+        # Create the stripe Customer, by default subscriber Model is User,
+        # this can be overridden with settings.DJSTRIPE_SUBSCRIBER_MODEL
+        customer, created = djstripe.models.Customer.get_or_create(subscriber=self.request.user)
+
+        # Add the source as the customer's default card
+        customer.add_card(stripe_source)
+
+        # Using the Stripe API, create a subscription for this customer,
+        # using the customer's default payment source
+        stripe_subscription = stripe.Subscription.create(
+            customer=customer.id,
+            items=[{"plan": plan.id}],
+            collection_method="charge_automatically",
+            # tax_percent=15,
+            api_key=djstripe.settings.STRIPE_SECRET_KEY,
+        )
+
+        # Sync the Stripe API return data to the database,
+        # this way we don't need to wait for a webhook-triggered sync
+        subscription = djstripe.models.Subscription.sync_from_stripe_data(
+            stripe_subscription
+        )
+
+        self.request.subscription = subscription
+
+        return super().form_valid(form)
+
+    def get_success_url(self):
+        return reverse(
+            "subscriptions:success",
+            kwargs={"id": self.request.subscription.id},
+        )
+
+
+class SubscriptionsSuccessView(DashboardView):
+    template_name = "subscriptions/success.html"
+
+    queryset = djstripe.models.Subscription.objects.all()
+    slug_field = "id"
+    slug_url_kwarg = "id"
+    context_object_name = "subscription"
+
+
+# TODO: Maybe this needs to be a normal class? Maybe throw it above?
+# Might not be possible due to "requires_action" but could use REST API thing
+# TODO: Use crispy form
+def subscriptions_payment_intent_view_create(request: HttpRequest):
```

Comment:
@leigh-johnson this is the actual functional code. The above are more or less to be removed.
How I can write this better? more elegant? I was thinking about REST framework that is used elsewhere in the project, is that the "right" way?
If yes, I'm a bit puzzled with the serializers. From the way I think of it, I don't need one, am I right?

(ignore absurd code such as `Plan.objects.first()` for now)

Reply:
Would it be easier to forget the idea of opening a modal, and instead render separate form/detail views? 
https://github.com/dj-stripe/dj-stripe/blob/stable/2.4/tests/apps/example/views.py

Reply:
I think the logic here is pretty contained to the webapp, so I don't think it makes sense to expose to the outside world via REST API. A template view makes sense to me. :+1: 

A good litmus test for "does this functionality belong in the REST API?" is:

* Do I need to `pip install print-nanny-client` and interact with REST endpoints? 
* Do I need to `npm install print-nanny-client` and interact with the REST endpoints?

Since subscriptions are contained within the webapp, there isn't a good reason to add REST interfaces on top of that. 

Does this distinction make sense?

Reply:
after many hours over that `views.py`, I understood that actually everything is irrelevant apart from the `template_file` inside `PurchaseSubscriptionView` and the `create_payment_intent` function. The rest of the code is for the old way of sources or tokens for stripe - which we don't want, as it does not support 3D secure, etc. and is incompatible with European law, from what I got.

While I agree with you, why would someone use the REST API for that, we need to transfer the `payment_intent_id` value inside the same step.

Maybe we could split the steps in 3-4 forms & views, but I'm not sure and I think that it's much more work rather than just keeping the current ugly code.

BTW the modal is now implemented and it's looking awesome

Reply:
Gotcha, thank you for digging deep into this and understanding what's needed! 🙇‍♀️

No worries about the first pass feeling inelegant. I'm always cool with a "make it work, then make it work more-better" approach. How're you feeling about edge case coverage? Are there any requirements remaining? 

:shipit: One quick way to deal with edge cases is to trace the request handler with Honeycomb. I think you can do something like...

I use tracing as a substitute for unit tests sometimes (or as a first step to writing them - study how it breaks in the wild)
```
import beeline

# A trace for this fn should be started automatically by HoneyComb middleware: https://github.com/bitsy-ai/octoprint-nanny-webapp/blob/main/print_nanny_webapp/middleware/honeycomb.py#L7

def subscriptions_payment_intent_view_create(request: HttpRequest):
    span = beeline.start_span(context={"request", request, "name": "parse_payment_intent_request" })
    

    if request.method != "POST":
        ctx = {"STRIPE_PUBLIC_KEY": djstripe.settings.STRIPE_PUBLIC_KEY}
        return TemplateResponse(request, SubscriptionsListView.template_name, ctx)

    intent = None
    data = json.loads(request.body)
    # TODO: Put this in the form
    data["plan"] = djstripe.models.Plan.objects.first().id
    beeline.finish_span(span)

    span = beeline.start_span(context={"request", request, "name": "some_other_payment_intent_logic" })
    # ... some other code  
    beeline.finish_span(span)

```

Reply:
No offense intended, by the way! I have a lot of confidence in your software writing skills. 

I basically assume all code (including / ESPECIALLY my own) will break in colorful and surprising ways 😂 ¯\_(ツ)_/¯

Reply:
ESPECIALLY about payments, lets not get reassured by my potentially "majestic" software skills and lets see those tests pass (thank you for your kind words though! :) )